### ISSUE TITLE
[Truffle] Error calling block in optparse for MRI tests.

### ISSUE THREAD

**Issue by bjfish** [jruby/jruby#2823]
Currently, blocking running the MRI tests.

Example:
```ruby
class ExampleHash < Hash
 
    def self.regexp(key, icase)
      Regexp.new('\A' + Regexp.quote(key).gsub(/\w+\b/, '\&\w*'), icase)
    end
 
    def self.candidate(key, icase = false, pat = nil, &block)
      pat ||= ExampleHash.regexp(key, icase)
      candidates = []
      puts "block #{block}"
      block.call do |k, *v|
        puts "INBLOCK"
        (if Regexp === k
           kn = nil
           k === key
         else
           kn = defined?(k.id2name) ? k.id2name : k
           pat === kn
         end) or next
        v << k if v.empty?
        candidates << [k, v, kn]
      end
      candidates
    end

    def candidates(key, icase = false, pat = nil)
      ExampleHash.candidate(key, icase, pat, &method(:each))
    end
    
    def test
      candidates( "no", false, /(?-mix:\Ano\w*)/ , &method(:each) )
    end

end

puts "`#{ExampleHash[ :no ,"no" ].test}`"
```
Expected
```
block #<Proc:0x007fced98e8fa0 (lambda)>
INBLOCK
`[[:no, ["no"], "no"]]`
```
Actual
```
block #<Proc:0xe0@core:/core/shims.rb:121>
`[]`

```

**Comment by bjfish** [comment]
The change in optparse.rb can be used to workaround this issue to run `jt test mri` tests:
https://gist.github.com/bjfish/1be84623da4a746f4d16

**Comment by chrisseaton** [comment]
Simpler case:

```
def bar
  yield 14
end

def foo(&block)
  block.call do |x|
    puts x
  end
end

foo &method(:bar)
```

```
$ ruby ../test.rb 
14
$ jt run ../test.rb 
../test.rb:2:in `bar': no block given (yield) (LocalJumpError)
  from core:/core/shims.rb:121:in `call'
  from core:/core/shims.rb:121:in `block in to_proc'
  from ../test.rb:6:in `call'
  from ../test.rb:6:in `foo'
  from ../test.rb:11:in `<main>'
```

Conversely:

```
def foo(&block)
  block.call do |x|
    puts x
  end
end

foo {
  yield 14
}
```

```
$ ruby ../test.rb 
../test.rb:8:in `block in <main>': no block given (yield) (LocalJumpError)
	from ../test.rb:2:in `call'
	from ../test.rb:2:in `foo'
	from ../test.rb:7:in `<main>'
$ jt run ../test.rb 
14
```

It looks like this is to do with what version of the method we use when something is a method and when it's a block - we store multiple versions of methods depending on how they're used.

**Comment by chrisseaton** [comment]
Fixed in 06c5d294600e5823bf22d1d4565063a722fdddfb
